### PR TITLE
Tracing: Add support to configure trace to logs start and end time

### LINKED ISSUE #31403: Improve Traces -> Logs Configuration
**Why is this needed**:
Currently, when building the Loki time range for the Loki query, the traces to logs link adds a harcoded amount of time to the end of the span.  The idea was to capture a little time after the span ended for context.  However, for services that log very rapidly this amount of time could easily push relevant logs off of the first page of Loki results.

**What would you like to be added**:
Make the amount of time after the end of a span configurable.  Default to 0.  i.e. the end time of the Loki query should match exactly to the end time of the span.

@slim-bean

### PR DESCRIPTION
**Pull request by zoltanbedi**
This PR adds support to set a custom amount time to shift for the trace to logs feature. Now the default behaviour is that it uses the spans start time and end time for the Loki query (it used to shift it by 1 hour to the past).

### Major change

Loki's range query is now uses milliseconds precision.

## TODO

- [x] Add docs

**Which issue(s) this PR fixes**:

Fixes #31403

### CODE REVIEW

**Comment by oddlittlebird** [comment]
Docs look good, but please make sure to update What's new in Grafana v8.0 so our users know about these changes.

**Review by achatterjee-grafana — APPROVED** [review]

File: public/app/core/components/TraceToLogsSettings.tsx [hunk 1]
```diff
@@ -6,12 +6,14 @@ import {
...
 import { DataSourcePicker } from '@grafana/runtime';
-import { InlineField, InlineFieldRow, TagsInput, useStyles } from '@grafana/ui';
+import { InlineField, InlineFieldRow, Input, TagsInput, useStyles } from '@grafana/ui';
 import React from 'react';
 
 export interface TraceToLogsOptions {
   datasourceUid?: string;
   tags?: string[];
+  spanStartTimeShift?: number;
+  spanEndTimeShift?: number;
```
Comment (aocenas) [inline public/app/core/components/TraceToLogsSettings.tsx:16]: Can we instead use the same code that is used for other time inputs? That can accept strings like `1s` or `5m`? Probably with defaulting to milliseconds if there is no unit. To align it with other time inputs.

Comment (zoltanbedi) [reply public/app/core/components/TraceToLogsSettings.tsx:16]: Good idea!

**Comment by aocenas** [comment]
Works fine but there are 2 things I am not sure about:
1. The config adds the time in different directions, ie extending to the past or to the future so adding -interval or +interval. Not sure if it is just me but it wasn't really clear to me. Maybe we could change the labels a bit to say, it extends to the past vs extends to the future.
2. Seems like we don't really do less then a second time ranges. For millisecond traces this means we jest use time range like 12:50:10 - 12:50:10. Also it means shifting by less than a second does not have much of an effect. Can you check if there is a way to use smaller than a second ranges?

**Comment by zoltanbedi** [comment]
@aocenas Added the possibility to use negative numbers and now Loki query range uses milliseconds precision.

File: docs/sources/datasources/jaeger.md [hunk 2]
```diff
@@ -32,8 +32,10 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
 
 - **Data source -** Target data source.
 - **Tags -** The tags that will be used in the Loki query. Default is `'cluster', 'hostname', 'namespace', 'pod'`.
+- **Span start time shift -** The time to shift the start time for the Loki query based on the span start time. In order to extend to the past you need to use negative value.  Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
+- **Span end time shift -** The time to shift the end time for the Loki query based on the span end time. Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/jaeger.md:36]:
```suggestion
- **Span start time shift -** Shift in the start time for the Loki query based on the span start time. In order to extend to the past, you need to use a negative value.  Time units can be used here, for example, 5s, 1m, 3h. The default is 0.
- **Span end time shift -** Shift in the end time for the Loki query based on the span end time. Time units can be used here, for example, 5s, 1m, 3h. The default is 0.
```

File: docs/sources/datasources/tempo.md [hunk 3]
```diff
@@ -31,8 +31,10 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
 
 - **Data source -** Target data source.
 - **Tags -** The tags that will be used in the Loki query. Default is `'cluster', 'hostname', 'namespace', 'pod'`.
+- **Span start time shift -** The time to shift the start time for the Loki query based on the span start time. In order to extend to the past you need to use negative value.  Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
+- **Span end time shift -** The time to shift the end time for the Loki query based on the span end time. Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/tempo.md:35]: Same doc change as above.

File: docs/sources/datasources/zipkin.md [hunk 4]
```diff
@@ -32,8 +32,10 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
 
 - **Data source -** Target data source.
 - **Tags -** The tags that will be used in the Loki query. Default is `'cluster', 'hostname', 'namespace', 'pod'`.
+- **Span start time shift -** The time to shift the start time for the Loki query based on the span start time. In order to extend to the past you need to use negative value.  Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
+- **Span end time shift -** The time to shift the end time for the Loki query based on the span end time. Time units can be used here, for example: 5s, 1m, 3h. The default is 0.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/zipkin.md:36]: Same doc change as above.

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc change request.

**Review by achatterjee-grafana — APPROVED** [review]
LGTM

File: public/app/plugins/datasource/loki/datasource.ts [hunk 5]
```diff
@@ -175,14 +175,10 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
         this.adjustInterval((options as DataQueryRequest<LokiQuery>).intervalMs || 1000, rangeMs) / 1000;
       // We want to ceil to 3 decimal places
       const step = Math.ceil(adjustedInterval * 1000) / 1000;
-      const alignedTimes = {
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/loki/datasource.ts:178]: This is the place where we rounded up the range to second precision.

File: public/app/features/explore/TraceView/createSpanLink.tsx [hunk 6]
```diff
@@ -79,25 +79,32 @@ function getLokiQueryFromSpan(span: TraceSpan, keys?: string[]): string {
...
+  const from = dateTime(adjustedStartTime);
   const spanEndMs = (span.startTime + span.duration) / 1000;
-  const to = dateTime(spanEndMs + 5 * 1000);
+  let adjustedEndTime = traceToLogsOptions?.spanEndTimeShift
+    ? spanEndMs + rangeUtil.intervalToMs(traceToLogsOptions.spanEndTimeShift)
+    : spanEndMs;
+
+  // Because we can only pass milliseconds in the url we need to round it and check if they equal.
+  // We need end time to be later than start time
+  if (Math.floor(adjustedStartTime) === Math.floor(adjustedEndTime)) {
```
Comment (zoltanbedi) [inline public/app/features/explore/TraceView/createSpanLink.tsx:96]: Would it make sense to move this check to the loki data source instead?

Comment (aocenas) [reply public/app/features/explore/TraceView/createSpanLink.tsx:96]: I think it make sense here. Maybe we can just do the `Math.floor(adjustedStartTime)` from the start where `adjustedStartTime` is created? I mean we won't use the fractional part anyway right?

**Review by aocenas — APPROVED** [review]
Looks good, nice work

File: public/app/features/explore/TraceView/createSpanLink.tsx (hunk 6 above)
Comment (zoltanbedi) [reply public/app/features/explore/TraceView/createSpanLink.tsx:96]: yes thats right

**Pull request merged**